### PR TITLE
feat: Redesigning/refactoring-the-digest-code-#2727

### LINKED ISSUE #2727: [NV-1649] ✨ Polishing: Redesigning/refactoring the digest code
### 📜 Description

While working on digest bug #2697  I felt the current digest code is too complex and excessive and can be simplified significantly, also performance also can be improved by refactoring and reducing the DB calls. I will create a draft PR and we can discuss in detail.

### 👟 Reproduction steps

_No response_

### 👍 Expected behavior

_No response_

### 📃 Provide any additional context for the Bug.

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/novuhq/novu/blob/main/CONTRIBUTING.md)

<sub>[NV-1649](https://linear.app/novu/issue/NV-1649/✨-polishing-redesigningrefactoring-the-digest-code)</sub>

### PR DESCRIPTION
**Pull request by tsssdev**
### What change does this PR introduce?

As part of this draft PR, mainly refactored digest code and also refactored the workflow queue code and fixed concurrent issues with out any complex code/locks. Most of the changes in this PR is related to old code cleanup/removal, new code mainly in the ProcessNotofication.ts and DigestService and constructDAG in helpers.ts and few other existing files.

In the old code logic is fragmented and spread many places, and digest decisions made  in different places in different conditions, now that is simplified and at the start itself creating DAG for complete execution flow and re-arranged nodes accordingly(which  steps runs parallel and which steps runs sequential decided at the start itself).

And removed trigger dummy jobs completely and delayed jobs removed and added delay to the next active node in the chain.. I personally feel delay should go with regular jobs/steps, take any queue system, or calendars(google etc.) the delay is part of events/jobs, so users also assumes that way I assume, and having separate delay node does not give any edge to user, in fact it  clutters UI when users have more jobs, alos later if we provide one consolidated graph of al emails configured then easily we may get 100 to  200 different emails jobs even for small to medium organization,  so there again adding delay nodes clutters it, instead of that we can have badge/ribbon on delay job, or different color.. so with out cluttering we can communicate which are delayed jobs.. anyhow this discussion we can do separately, as I did not modify UI, for UI we still have delay jobs, in the backend while scheduling avoiding unnecessary delay jobs and adding delay to real jobs.. 

Here are the assumptions made:(I could not find any doc explains all possible cases)
1. If we get more than one digest jobs in the chain then all the jobs runs parallel, say we have  d1, s2, d3, s4, s5, s6 in the chain then this forms two parallel chains d1, s2 and 3, s4, s5, s6 , like this how many digest nodes those many parallel chains.
2. If delay node has delaypath in it then it forms another top level chain because delaypath means we do not when those notifications triggered, these are controlled by the payload..
3. If digest job is inactive then complete chain will be inactive, if normal node inactive then the nodes below also becomes inactive, is this OK?

These assumptions are just to start, if we need to modify we can modify based on feedback, in the new code it is very easy to change what ever our requirement is, just changing DAG enough most cases. 

### Why was this change needed?

Closes #2727

### CODE REVIEW

File: libs/shared/src/types/channel/index.ts [hunk 1]
```diff
@@ -13,7 +13,6 @@ export enum StepTypeEnum {
   CHAT = 'chat',
   PUSH = 'push',
   DIGEST = 'digest',
-  TRIGGER = 'trigger',
```
Comment (p-fernandez) [inline libs/shared/src/types/channel/index.ts:16]: What's the reason to remove the TRIGGER step?

File: packages/application-generic/src/factories/sms/handlers/base.handler.ts [hunk 2]
```diff
@@ -21,9 +21,7 @@ export abstract class BaseSmsHandler implements ISmsHandler {
 
   async send(options: ISmsOptions) {
     if (process.env.NODE_ENV === 'test') {
-      throw new Error(
-        'Currently 3rd-party packages test are not support on test env'
-      );
+      return {};
```
Comment (p-fernandez) [inline packages/application-generic/src/factories/sms/handlers/base.handler.ts:24]: These changes on the handlers seem very arbitrary and without a clear purpose.

File: apps/api/src/app/events/helpers/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    dag[dag.length - 1].push({ ...step, delay: getMetaDelay(step) });
+
+    return dag;
+  }, [] as StepWithDelay[][]);
+
+  return resultDag.map((branch) => {
+    const inactiveIndex = branch.findIndex((step) => !step.active);
+
+    return inactiveIndex === -1 ? branch : branch.slice(0, inactiveIndex);
+  });
```
Comment (p-fernandez) [inline apps/api/src/app/events/helpers/helpers.ts:28]: It is hard to follow up the code and I don't even know what DAG stands for (acronyms without any context nor documentation are not very helpful unfortunately) but I guess this is the place where the new main logic starts, is that right? 
My surface understanding is that here you create it like a list in order of the different jobs a triggered event needs to execute. If that's right, it is an idea we were considering to solve some of the problems we were having: to create all the jobs needed to execute before queuing and then change the status or cancel them based on the other functionalities we have before executing them from the queue. That would help to remove many calculations and the locks we had to introduce to assure right behaviour against concurrency (at least for the regular Digest; Backoff digest still not working properly; I neither found here that you fully fixed it).

One of the problems with this implementation is to consider if in the future we want to implement conditional workflows (think in decision trees) how flexible a linear creation before queueing would be. So that would need time to think about it.

**Review by p-fernandez — COMMENTED** [review]
As always thanks for your contribution and your big efforts proposing improvements and fixes. I didn't get too deep into the formality and implementation details as I think there are bigger things to discuss first.

You have considered a direction in the job and steps generation for an event that we have discussed to try to solve many of the problems, and that's a good thing. I extend myself in the relevant comment about it. If we finally decide going for it, your work could be helpful.
Better explanations and/or diagrams of what your design is would have been quite helpful for us to understand the proposal better.

Unfortunately this PR has a fair amount of problems:
- it is hard to follow and understand due to size, amount of problems tackled at the same time in one single PR, clarity of code (it doesn't follow our code architecture in many areas or just moves things around without a clear purpose) and arbitrary decisions (totally understandable coming from a contributor that is not 100% aware of the directions discussed in the core team).

- the implementation and changes are not retro-compatible (by choice) with the current system and that's something we can't skip as we are responsible of providing a functional product to our users

- it breaks existing features such the Activity Feed and Advanced Debugging (Execution Details)

- it removes the Delay step functionality that's totally different to the Digest functionality

- and mainly it doesn't fix the current concurrency problem we have with the backoff digest functionality. I tried it with some complex workflow scripts that I have to test manually implementations and it is not working fully right unfortunately (neither in the current status of `next` and it is work ongoing). 

My suggestion, as I did the Providers proposal you presented (https://github.com/novuhq/novu/pull/2647), is to split this PR in multiple different PRs with pieces of work that can be reviewed easily and can be implemented iteratively keeping the compatibility with the current system, so far I would see:
- Fixing Backoff Digest concurrency problem with the current implementation we have (we are already working on it, but it is not an easy task)

- The abstraction of WorkflowQueue dependency and its service, removing all domain logic from them and all DAL dependencies (something that I also already noticed as you could see from other PRs: https://github.com/novuhq/novu/pull/2919)

- Those 2 would be needed to be able to do the rewrite of the trigger event implementation and its interaction with the queue.

Also when changing so many moving parts it should be mandatory to add more thorough testing to the parts changed (and if tests are missing), as that would have raised some of the things I noticed this PR breaks. Not only to test the new functionalities implemented.

CC: @scopsy

File: libs/dal/src/repositories/notification/notification.schema.ts [hunk 4]
```diff
@@ -34,9 +34,6 @@ const notificationSchema = new Schema<NotificationDBModel>(
         type: Schema.Types.String,
       },
     ],
-    _digestedNotificationId: {
-      type: Schema.Types.String,
-    },
```
[the remark this answers is not in the document]
Comment (tsssdev) [reply libs/dal/src/repositories/notification/notification.schema.ts:39]: @p-fernandez  Thank you for the detailed feedback, I will reply one by one... coming to DAG, it is Directed acyclic graph, from sequence of steps creating a parallel execution graph, example, all digest steps should start parallel, it does not make sense to run those sequentially, delay steps with delaypath also should run parallel, sequence does not make sense for those jobs as well, because we do not know what users pass for the delaypath.  Coming to trigger jobs, those are unnecessary in this new design.. and delay jobs alos I felt unnecessary, delay should be an attribute of the job/event, look at any queue system/calendar, the delay is always associated with the job/event, that is why I felt it is unnecessary and propagated delay to next active step. I believe for end users also having delay part of jobs is more comfortable and intuitive(probably take a poll or feedback?), but in any case if we keep delay on the front end also the api code works fine, with out breaking anything we just skipping delay jobs behind the scenes. Also can you let me know which cases it is not working?  did you try this on latest code that has digestRunKey? I tried all possible ways and it is working fine for me. If you can share the script I can try it and check which cases failing.. I also uncommented some of the old test cases commented in old code and working fine in new code and also created parallel dynamic generation of test cases by running parallel execution of 300 events with multiple subscribers. if you can share the script then I can try and fix it if any issues there.. As it is draft PR, I was just focusing on the API changes, if this direction is fine, I can create final PR and that time I can have complete impact analysis and fix other parts of the code as well, also looks like lot of changes happening on the next branch side, I have adjusted some code, but couple of cases skipped it(will discuss more on this later), yes backward compatibility I totally agree, I believe all these changes we can do with out breaking old functionality, even details logs also mostly taken care, if any details logs missing we can add easily in final PR.
Here are the specs followed in this PR:
1. Digest and delay with path always at the root level of DAG
2. Delays are added to next active step and delay steps removed
3. Two consecutive delays adds up and added to next active step
4. Delays just before digest are ignored
5. Delays at the leaf level of branches ignored
6. Steps before digest forms a parallel branch
7. Delay path step forms parallel branch
8. two consecutive digest steps forms two branches, one branch has just digest node
9. Negative delays in meta should throw error
10. Negative/missing delays in payload and overrides should ignore and proceed with delay 0
11. Missing delay path in meta should throw error for delay nodes
12. If any inactive step present in the branch then all the steps below will be trimmed

Some of these specs based on current code/docs and some places where we don't have clear doc in existing system, made up my own, If we need to change any of these please let me know, we can adjust code very easily

**Comment by tsssdev** [comment]
@p-fernandez Thank you for the detailed feedback, I will reply one by one... coming to DAG, it is Directed acyclic graph, from sequence of steps creating a parallel execution graph, example, all digest steps should start parallel, it does not make sense to run those sequentially, delay steps with delaypath also should run parallel, sequence does not make sense for those jobs as well, because we do not know what users pass for the delaypath. Coming to trigger jobs, those are unnecessary in this new design.. and delay jobs alos I felt unnecessary, delay should be an attribute of the job/event, look at any queue system/calendar, the delay is always associated with the job/event, that is why I felt it is unnecessary and propagated delay to next active step. I believe for end users also having delay part of jobs is more comfortable and intuitive(probably take a poll or feedback?), but in any case if we keep delay on the front end also the api code works fine, with out breaking anything we just skipping delay jobs behind the scenes. Also can you let me know which cases it is not working? did you try this on latest code that has digestRunKey? I tried all possible ways and it is working fine for me. If you can share the script I can try it and check which cases failing.. I also uncommented some of the old test cases commented in old code and working fine in new code and also created parallel dynamic generation of test cases by running parallel execution of 300 events with multiple subscribers. if you can share the script then I can try and fix it if any issues there.. As it is draft PR, I was just focusing on the API changes, if this direction is fine, I can create final PR and that time I can have complete impact analysis and fix other parts of the code as well, also looks like lot of changes happening on the next branch side, I have adjusted some code, but couple of cases skipped it(will discuss more on this later), yes backward compatibility I totally agree, I believe all these changes we can do with out breaking old functionality, even details logs also mostly taken care, if any details logs missing we can add easily in final PR.
Here are the specs followed in this PR:

    Digest and delay with path always at the root level of DAG
    Delays are added to next active step and delay steps removed
    Two consecutive delays adds up and added to next active step
    Delays just before digest are ignored
    Delays at the leaf level of branches ignored
    Steps before digest forms a parallel branch
    Delay path step forms parallel branch
    two consecutive digest steps forms two branches, one branch has just digest node
    Negative delays in meta should throw error
    Negative/missing delays in payload and overrides should ignore and proceed with delay 0
    Missing delay path in meta should throw error for delay nodes
    If any inactive step present in the branch then all the steps below will be trimmed

Some of these specs based on current code/docs and some places where we don't have clear doc in existing system, made up my own, If we need to change any of these please let me know, we can adjust code very easily

File: libs/dal/src/repositories/notification/notification.schema.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tsssdev) [reply libs/dal/src/repositories/notification/notification.schema.ts:39]: Sure In final PR we can adjust this code easily, instead of _digestedNotificationId we just need to use the _digestedNotificationIds array and we can remove _digestedNotificationId  column., small change.

File: libs/dal/src/repositories/notification-template/notification-template.entity.ts [hunk 5]
```diff
@@ -45,8 +45,6 @@ export class NotificationTemplateEntity {
 
   _notificationGroupId: string;
 
-  _parentId?: string;
```
[the remark this answers is not in the document]
Comment (tsssdev) [reply libs/dal/src/repositories/notification-template/notification-template.entity.ts:48]: I have gone through entire code and all places steps are used like chain/array and removing this has zero impact, but in cases if you guys think we need to put back we can put back, this has zero impact in new code.

File: apps/web/src/pages/activities/hooks/useNotificationStatus.ts [hunk 6]
```diff
@@ -13,7 +13,7 @@ export const useNotificationStatus = (notification: any): JobStatusEnum => {
         if (prev === JobStatusEnum.FAILED) {
           return prev;
         }
-        if ([JobStatusEnum.PENDING, JobStatusEnum.DELAYED].includes(prev)) {
+        if ([JobStatusEnum.PENDING, JobStatusEnum.QUEUED].includes(prev)) {
```
[the remark this answers is not in the document]
Comment (tsssdev) [reply apps/web/src/pages/activities/hooks/useNotificationStatus.ts:16]: Yes this may include regular jobs sometimes briefly,  was just in two minds.(one way is it shows all queued jobs at that moment so we can see all jobs, other way is why to show regular queued jobs even for short times), if needed we can add delay > 0 condition in the final PR

File: apps/api/src/app/events/usecases/cancel-delayed/cancel-delayed.usecase.ts [hunk 7]
```diff
@@ -1,33 +1,32 @@
...
-  constructor(private jobRepository: JobRepository) {}
+  constructor(private jobRepository: JobRepository, private workflowQueueService: WorkflowQueueService) {}
 
   public async execute(command: CancelDelayedCommand): Promise<boolean> {
     const job = await this.jobRepository.findOne({
       _environmentId: command.environmentId,
       transactionId: command.transactionId,
-      status: JobStatusEnum.DELAYED,
+      status: JobStatusEnum.QUEUED,
+      delay: { $gt: 0 },
```
[the remark this answers is not in the document]
Comment (tsssdev) [reply apps/api/src/app/events/usecases/cancel-delayed/cancel-delayed.usecase.ts:14]: My thinking is delay should be part of the step/job(like an attribute), you take up any queue system/calendar, delay is part of event/job, there will not be any separate job for delay. I believe majority of our user perception also same.. Internally if we need to have rigid check we can move this to metadata(I have some ideas how to consolidate digest and delays, we can discuss later), coming to retrocompatible, this is fine with version upgrades, as part of this PR we are not deleting the delay jobs from DB, in DB still we have delay jobs like old way, while running we are merging delay jobs in to next active job and adding this delay in to next active job, so it works same like before(only in detail execution log instead of two steps we will have only one entry)

File: apps/api/src/app/events/services/digest/digest-service.ts [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+  ): Promise<string | undefined> {
+    const channelType = STEP_TYPE_TO_CHANNEL_TYPE.get(stepType);
+    if (!channelType) return;
+    const integrations = await this.getDecryptedIntegrations.execute(
+      GetDecryptedIntegrationsCommand.create({ channelType, active: true, organizationId, environmentId, userId })
+    );
+    const integration = integrations[0];
+
+    return integration?.providerId ?? InAppProviderIdEnum.Novu;
+  }
```
[the remark this answers is not in the document]
Comment (tsssdev) [reply apps/api/src/app/events/services/digest/digest-service.ts:189]: Yes, this is not related to redesign(that is why added comment: //for now keep it here), this is existing code and  need to use two places, so temporarily I kept it here, I could not find any common utils/service type of thing,  may be I can create one more file and import that in process notification and here? or I can move this to helpers.js? if you have any other common place, plesae let me know, I can move there, it is small change.

File: libs/shared/src/types/channel/index.ts (hunk 1 above)
Comment (tsssdev) [reply libs/shared/src/types/channel/index.ts:16]: We don't need these dummy jobs in new design, all dummy trigger jobs gone. The payload accumulation happening with _digestedNotificationIds of digest job.

File: packages/application-generic/src/factories/sms/handlers/base.handler.ts (hunk 2 above)
Comment (tsssdev) [reply packages/application-generic/src/factories/sms/handlers/base.handler.ts:24]: Yes, this is not related to redesign, but when executing test cases, it is trowing exceptions and confusing, so as in email, I just changed it, so we can avoid these useless exceptions..  BTW one more problem is right now all test cases steps are created with shouldStopOnFail=false, so that means almost all test cases tested for failed steps only, for success flow never tested. anyhow if you think this causes any problem I can put back that exception throw logic.

File: apps/api/src/app/events/services/workflow-queue/workflow-queue.ts [hunk 9]
```diff
@@ -0,0 +1,77 @@
+import { JobsOptions, Queue, QueueBaseOptions, QueueScheduler, Worker } from 'bullmq';
+import { getRedisPrefix } from '@novu/shared';
+import { ConnectionOptions } from 'tls';
+import { JobEntity, JobStatusEnum } from '@novu/dal';
+
+export type MinimalJob = Pick<JobEntity, '_id' | '_environmentId' | '_organizationId' | 'type' | 'delay' | '_userId'>;
+
+export class WorkflowQueue {
```
[the remark this answers is not in the document]
Comment (tsssdev) [reply apps/api/src/app/events/services/workflow-queue/workflow-queue.ts:8]: You are referring the JobEntity ? This MinimalJob export can be moved to shared/common place and imported all places including in workflow-queue.ts, it is easy change, I can do in final PR.. OR you are referring some other logic?

File: apps/api/src/app/events/usecases/cancel-delayed/cancel-delayed.usecase.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (SebastianStehle) [reply apps/api/src/app/events/usecases/cancel-delayed/cancel-delayed.usecase.ts:14]: Is a delay not a digest with a random digest key?

File: apps/api/src/app/events/helpers/helpers.ts (hunk 3 above)
Comment (tsssdev) [reply apps/api/src/app/events/helpers/helpers.ts:28]: DAG means Directed Acyclic Graph, to make it method name short I used DAG, coming to functionality we are taking the linear chain and constructing a gragh with the parallel paths, lets say if we have SMS1->Email1->Digest1->Email2->Digest2-Chat1 then it will create a parallel paths: 
1. SMS1->Email1
2. Digest1-> Email2
3. Digest2 -> Chat1

All three paths can executes parallel, and coming to jobs creation, no need to create upfront, that is our wish, when ever needed we can construct these DAGs, if needed we can cache the basic DAG along with template(when we have big chains caching may help), please look at spec tests file https://github.com/tsssdev/novu/blob/d058a888b28a144bb46956f4eb9847d772d553c3/apps/api/src/app/events/helpers/helper.dag.spec.ts for better understanding.
Coming to use cases, what ever use cases I can think off, the new code can handle very easily.. here are few use cases:
1. Multiple digest nodes(this is already implemented in new code)
2. Parallel jobs, instead of one sequential chain (already support is builtin in new code)
3. Daily/weekly digests at specified time, easy
4. deliver messages at the subscriber time zone.. in this new code easy
5. Count based digest, instead of time based if they want to deliver for every 1000(batch), it is very easy to implement in new code..
6. If they want to mix with count and time-based digest(which ever comes first), again easy in new code
7. Backoff after sending 10 messages, again easy.
8. Backoff count and timebased, easy.
9. cron/rrule support, easy

Throw me a use case you think tough then I can try to show case how easy to implement in this new code.

One more thing is even though I started with digest refactoring, I refactored complete jobs code, so not only digest jobs, even regular jobs benefit with these changes. and delay and trigger jobs overhead is gone(functionality not effected)

**Comment by tsssdev** [comment]
>     * Fixing Backoff Digest concurrency problem with the current implementation we have (we are already working on it, but it is not an easy task)

I did not focus much on back off logic, but I think it is easy to fix in new design, let me try quickly.

> 
>     * The abstraction of WorkflowQueue dependency and its service, removing all domain logic from them and all DAL dependencies (something that I also already noticed as you could see from other PRs: #2919)
> 
>     * Those 2 would be needed to be able to do the rewrite of the trigger event implementation and its interaction with the queue.
> 
Coming to workflow queue, already queue logic is moved to new ts file and old file now has some control logic, no queue logic, what other changes we need?

**Comment by tsssdev** [comment]
> > ```
> > * Fixing Backoff Digest concurrency problem with the current implementation we have (we are already working on it, but it is not an easy task)
> > ```
> 
> I did not focus much on back off logic, but I think it is easy to fix in new design, let me try quickly.
> 
This is fixed now, juts two files, few lines changed, now back-off working perfectly with out any issues(tested by generating parallel load with small script), can you cross check once with your scripts and see if you find any issues? Here also followed same technique what is followed for digest concurrency.(UUIDV5 with digestRunKey)

File: apps/api/src/app/events/usecases/cancel-delayed/cancel-delayed.usecase.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (tsssdev) [reply apps/api/src/app/events/usecases/cancel-delayed/cancel-delayed.usecase.ts:14]: @SebastianStehle based on our current assumption of digest job, delay job is different, but you are right, delay is like a digest job that will have unique random digest key and always digest one message, I consider even regular jobs are digest jobs with 1 message digested, I don't see why why we need to have different types of logic, delays and digests can be attributes of any step/job. if users does not enable digest then it becomes regular job(with digest count 1), same with delay...

**Comment by tsssdev** [comment]
Even though the PR looks big(78 files) the actual logic in new code is small, most of the changes in the PR is removal of old digest/delay/trigger code and testcases related(all the code is inter linked so I had to remove to make decent code instead of doing patches), all these changes are backward compatible, but if any cases missed I can fix in final PR, I was just focusing to show case this new approach. 

Coming to the logic, we need to focus on these  below 6 files:
For understanding DAG and delay merge logic(This is very important)
apps/api/src/app/events/helpers/helper.dag.spec.ts
apps/api/src/app/events/helpers/helpers.ts

For Queue refactoring:
apps/api/src/app/events/services/workflow-queue/workflow-queue.ts
apps/api/src/app/events/services/workflow-queue/workflow.queue.service.ts

For Digest and other logic
apps/api/src/app/events/services/digest/digest-service.ts
apps/api/src/app/events/usecases/process-notification/process-notification.usecase.ts

Remaining all (runJob etc) small small changes, can explore when those files come across. 

One more important thing is uuid v5 and digestRunKey.

And I suggest to look at this new code with fresh mind, if you try to map with old code this may confuse as this new code is followed completely different approach and some cases I preferred service pattern instead of usecase command pattern to keep the code together with out unnecessary repetition. After understanding you can compare with old code.

File: libs/dal/src/repositories/notification-template/notification-template.entity.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ainouzgali) [reply libs/dal/src/repositories/notification-template/notification-template.entity.ts:48]: @tsssdev @p-fernandez it will have affect. The _parentId of a notification template is the id of template in the environment before. Meaning  a template in production will have _parentId equals to the id that template in development.

[the remark this answers is not in the document]
Comment (tsssdev) [reply libs/dal/src/repositories/notification-template/notification-template.entity.ts:48]: Sure @ainouzgali I can put that back.

**Comment by SebastianStehle** [comment]
I hope it is okay to ask a question here. Because there is one thing I do not really understand. I am not sure if you want to handle it here, but I think somehow you are already doing it.

The visual workflow representation implies that there is a sequential execution of the steps. So first the notification is sent by email and then by SMS. It is actually unclear whether this is actually sequential.

* If the workflow is sequential, a delay in step would slow other steps as well. For example I think you already have a retry strategy and a fallback strategy for other providers. The big question is: When is a job completed and do you want to wait for the job to be done or not? If it is sequential then you also want to configure timeouts, because I guess you might want to give up after like 10 minutes and move forward with another channel.

* If the workflow is actually executed in parallel it becomes complicated when you have conditions or things like a shared context. It is then not very transparent how the workflow really works.

As a user I see two potential options that would make sense for me:

1. Give everything to the user. Introduce additional blocks like branching and "parallel" and build a real workflow engine that is transparent (and can be simulated). This moves a lot of responsibility to the user but also freedom to build stuff himself.

2. Get rid of the workflow and use a simpler alternative. For me it seems that you only want independent branches that basically work like this:

a) Filter
b) Digest or Delay
c) Filter
d) Execute

**Comment by tsssdev** [comment]
@SebastianStehle I am not paralyzing all, lets say you have email, digest and sms, here email and digest go parallel, digest/sms does not depend on email, email success or failure, digest/sms continues, both are independent paths, same with delaypath, delaypath is controlled by users, so that is again independent and parallel. our regular delay nodes are sequential.

**Comment by tsssdev** [comment]
Hi Guys, any one able to find flaws or able to break the new code? or found a use case that is difficult to implement in new code? I am have tested many ways and working rock solid in all possible cases. I believe this code is much more stable than what we have on live and performance is exponentially faster when compare to old code what we have on live, meanwhile I have also applied required FE(web) changes  this new UI is has more functionality(date range selection,  bar and linear graphs, daily, weekly, monthly graphs, no digest/delay job pollution/cluttering) and more importantly this FE related code also much more faster (less DB queries and connections, less data transfer between Browser to API server and API server to DB and the queries are more efficient, even after adding all these additional functionality the new queries runs faster than what we have on live, so more functionality and less load on the servers.  Put back most of the backward compatibility code(if I miss any edge cases, can add it back easily) and right now only backward compatibility we need to take care is what to do with old data that has unnecessary trigger/digest/delay jobs,  we can write simple scripts  to migrate to new format, so we can remove the cluttering in old data as well... based on our requirements we can write these scripts... before spending more time I would like to see if of guys ready for these changes, so while creating the final PR I can take care of backward compatible edge cases and data migration scripts,  if you guys not ready for these changes, lets move on. I have prepared small video showing the new UI, please check: https://www.loom.com/share/d4f5462fb1da4f80bf1476da312d885d

**Comment by scopsy** [comment]
Hi @tsssdev, unfortunately, as stated multiple times and even of our desire to do so, we cannot merge a 3,000 lines change with over 100 files in a single PR. 

This PR needs to be broken down into small pull requests that tackle a **single** change, we can review test, and discuss before merging. At this moment, I have no way of merging this one into our repository.

I really much appreciate your help and effort, but we just need to make sure we are following a good development practice of working in small chunks of work we can properly review and ship.

**Comment by tsssdev** [comment]
@scopsy This is complete redesign of the workflow queue and job execution logic(mainly digest related code but other jobs also effected) and the approach followed is completely different than existing code, I do not think it is possible to have small PRs, if it is just refactoring existing code by creating wrapper use cases or move code here and there then small PRs possible.. but if we need to make meaning full enhancements we have to go with decent PR, as I told, new code is not very big, the most of the code changes are related to removing old code.. and frankly speaking the current digest code on live is not enterprise ready, if not now, later also need to address that...

**Comment by scopsy** [comment]
Following up our conversation working with smaller PR's, will close this one for now 🙏
